src/dftbp/timedep/linrespcommon.F90

```diff
@@ -758,119 +776,939 @@ subroutine actionAminusB(spin, wij, win, nocc_ud, nvir_ud, nxoo_ud, nxvv_ud, nxo
+    nLoc = locSize(iam+1)
+    @:ASSERT(nmat == nLoc)
+    iGlb = vOffset(iam+1) + 1 
+    fGlb = vOffset(iam+1) + nLoc
+    allocate(vLoc, mold=vin)
+    vLoc = 0.0_dp
+    vout = 0.0_dp
+
+    if(tAplusB) then
+      vLoc(:) = vin
+    else
+      vLoc(:) = vin * sqrt(wij(iGlb:fGlb))
+    endif
+
+    !> Compute w_ia = q^ia_A G_AB q^jb v_jb 
+
+    oTmp(:) = 0.0_dp    
+    do myia = 1, nLoc
+      ia = vOffset(iam+1) + myia
+      qij(:) = transChrg%qTransIA(ia, env, denseDesc, ovrXev, grndEigVecs, getIA, win)
+      otmp(:) = otmp(:) + qij(:) * vLoc(myia) * sqrOccIA(ia)
+    enddo
+
+    call mpifx_allreduceip(env%mpi%globalComm, otmp, MPI_SUM)
+
+    if (.not.spin) then !-----------spin-unpolarized systems--------------
+
+      if (sym == 'S') then
+
+        call hemv(gtmp, gamma, otmp)
+
+        ! 4 * wn * (g * Q)
+        vOut(:) = 0.0_dp
+        do myia = 1, nLoc
+          ia = vOffset(iam+1) + myia
+          qij(:) = transChrg%qTransIA(ia, env, denseDesc, ovrXev, grndEigVecs, getIA, win)
+          vOut(myia) = 4.0_dp * sqrOccIA(ia) * dot_product(qij, gTmp)
+        enddo  
+
+      else
+       
+        otmp = otmp * spinW(species0)
+
+        ! 4 * wn * (o * Q)
+        vOut = 0.0_dp
+        call transChrg%qVecMat(env, denseDesc, ovrXev, grndEigVecs, getIA, win, oTmp, vOut, vOffset(iam+1))
+        vOut(:) = 4.0_dp * sqrOccIA(iGlb:fGlb) * vOut
+
+
+      end if
+
+    else !--------------spin-polarized systems--------
+
+      call hemv(gtmp, gamma, otmp)
+
+      otmp(:) = 0.0_dp
+
+      !$OMP PARALLEL DO DEFAULT(SHARED) PRIVATE(ia,ss,qij) &
+      !$OMP& SCHEDULE(RUNTIME) REDUCTION(+:otmp)
+      do myia = 1, nLoc
+         
+        ia = vOffset(iam+1) + myia
+        qij(:) = transChrg%qTransIA(ia, env, denseDesc, ovrXev, grndEigVecs, getIA, win)
+
+        ! singlet gamma part (S)
+        vOut(myia) = 2.0_dp * sqrOccIA(ia) * dot_product(qij, gtmp)
+
+        ! magnetization part (T1)
+        ss = getIA(win(ia), 3)
+
+        otmp(:) = otmp + spinFactor(ss) * sqrOccIA(ia) * vLoc(myia) * qij
+
+      end do
+      !$OMP  END PARALLEL DO
+
+      call mpifx_allreduceip(env%mpi%globalComm, otmp, MPI_SUM)
+
+      otmp = otmp * spinW(species0)
+
+      !$OMP PARALLEL DO DEFAULT(SHARED) PRIVATE(ia,ss,qij) &
+      !$OMP& SCHEDULE(RUNTIME)
+      do myia = 1, nLoc
+
+        ia = vOffset(iam+1) + myia        
+        qij(:) = transChrg%qTransIA(ia, env, denseDesc, ovrXev, grndEigVecs, getIA, win)
+
+        ss = getIA(win(ia), 3)
+      
+        vOut(myia) = vOut(myia) + 2.0_dp * sqrOccIA(ia) * spinFactor(ss) * dot_product(qij, otmp)
+
+      end do
+      !$OMP  END PARALLEL DO
+
+    end if
+    
+    !> The largest matrices in this part are of dim nAtoms*nVir*nVir, seems logical to distribute 
+    !> them. The local dimensions are different from the case of nOcc*nVir, we require new offsets.
+    !> A new index array is required because the standard abs index is symmetrical in a and b  
```

Comment:
```suggestion
    ! The largest matrices in this part are of dim nAtoms*nVir*nVir, seems logical to distribute 
    ! them. The local dimensions are different from the case of nOcc*nVir, we require new offsets.
    ! A new index array is required because the standard abs index is symmetrical in a and b  
```